### ISSUE TITLE
Implement einsum

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#124]
Split conversation from https://github.com/mratsim/Arraymancer/issues/58. cc @szabi

**Comment by mratsim** [comment]
@szabi We can probably use Numpy/Tensorflow implementation as reference but `einsum` has several challenges:

- This is quite an involved function: see [Numpy implementation.](https://github.com/numpy/numpy/blob/v1.13.0/numpy/core/einsumfunc.py#L703-L993) and even 
Tensorflow does not fully implement it see [here](https://github.com/tensorflow/tensorflow/issues/4722).
- Licensing, we can't just lift it as is
- What format to use as argument (strings? which convention) and how to parse it
   - np.einsum('ii->i', a)
   - np.einsum('...j,j', a, b)
   - np.einsum('ki,...k->i...')
- Optimization (OpenMP, SIMD, ...)
- How to leverage
   - CuBLAS [StridedBatchedGEMM](https://devblogs.nvidia.com/parallelforall/cublas-strided-batched-matrix-multiply/), wrapped [here](https://github.com/mratsim/Arraymancer/blob/218aa7bb0b412903847b8761cf6c266c53fc300d/src/tensor/backend/cublas.nim#L200)
   - BLIS and the fallback strided GEMM
   - MKL batched GEMM
   to be fast enough and limit our manual implementation

Worth keeping track of:
- Very simple einsum: https://github.com/khundman/Einsum/blob/master/myeinsum.py
- Optimized einsum: https://github.com/dgasmith/opt_einsum
- Parallel einsum: https://jackkamm.github.io/blog/a-parallel-einsum/
- PyTorch WIP: https://github.com/pytorch/pytorch/pull/2008
- Tensorflow [implementation](https://github.com/tensorflow/tensorflow/blob/0ffefe178f71ddb595471ddfc43edb8d702155ca/tensorflow/python/ops/special_math_ops.py#L85)
- Julia [Tensor contractions](https://github.com/Jutho/TensorOperations.jl)
- Any "tensor contraction" resources

**Comment by mratsim** [comment]
I've pondered a bit on the subject.

Here is a drill down of [einsum various use cases](https://obilaniu6266h16.wordpress.com/2016/02/04/einstein-summation-in-numpy/). The examples can serve as the base test suite (comparing einsum vs manual loops).
Images from Olexa Bilaniuk's blog
![image](https://user-images.githubusercontent.com/22738317/31884755-788be626-b7ee-11e7-9d60-a579ac22bb55.png)
![image](https://user-images.githubusercontent.com/22738317/31884759-7c612cac-b7ee-11e7-9bda-123e4bbdd035.png)
![image](https://user-images.githubusercontent.com/22738317/31884633-1730a4b6-b7ee-11e7-8e10-087a880cfdd0.png)
![image](https://user-images.githubusercontent.com/22738317/31884639-1b6a453c-b7ee-11e7-888a-1395e0b2d9e4.png)
![image](https://user-images.githubusercontent.com/22738317/31884643-214e7e6e-b7ee-11e7-8106-d3c27c43cb12.png)
![image](https://user-images.githubusercontent.com/22738317/31884648-25b21236-b7ee-11e7-882e-3e7504c19203.png)
![image](https://user-images.githubusercontent.com/22738317/31884655-2abe4312-b7ee-11e7-85a5-8269173d1467.png)
![image](https://user-images.githubusercontent.com/22738317/31884813-acfe9bf6-b7ee-11e7-9d44-51c89edc63eb.png)
![image](https://user-images.githubusercontent.com/22738317/31884665-32834d40-b7ee-11e7-8a8a-6509464c7729.png)
![image](https://user-images.githubusercontent.com/22738317/31884675-38928b06-b7ee-11e7-8f43-171a51e2f32e.png)
![image](https://user-images.githubusercontent.com/22738317/31884684-3e2b3888-b7ee-11e7-91d2-aab58e264dd5.png)
![image](https://user-images.githubusercontent.com/22738317/31884689-41d3501a-b7ee-11e7-80ce-14189fa796c3.png)
![image](https://user-images.githubusercontent.com/22738317/31884693-473d54f6-b7ee-11e7-874a-959e1895cb8c.png)
![image](https://user-images.githubusercontent.com/22738317/31884708-5143c85e-b7ee-11e7-9a3a-c9fbd89e728a.png)
![image](https://user-images.githubusercontent.com/22738317/31884716-57e902be-b7ee-11e7-98ae-ea1b369b4bcf.png)
![image](https://user-images.githubusercontent.com/22738317/31884726-5ffb4c82-b7ee-11e7-8d79-9c3a785abbac.png)


From this, I think that we can implement something much more elegant than Python by metaprogramming the loops.

Resources:
- [Einsum implementation in Julia](https://github.com/ahwillia/Einsum.jl/blob/master/src/Einsum.jl) using metapgrogramming
- [Einsum Numpy C reimplemented in Python](https://github.com/hpaulj/numpy-einsum)
- [Port of Tensorflow einsum to PyTorch Python](https://gist.github.com/rockt/15ee013889d65342088e9260a377dc8f)
- [Fastor](https://github.com/romeric/Fastor), SIMD-based (AVX, etc) C++11 linear-algebra framework. They also have [einsum based on tensor contractions](https://github.com/romeric/Fastor/blob/7a4ad88f25fa23f2e32c478f04684ae8671e2e50/tensor_algebra/einsum.h)

GPU note: generating a GPU kernel will probably be hard for the cases not covered by cublas stridedbatchedGEMM

**Comment by szabi** [comment]
> * Licensing, we can't just lift it as is

As I see, Numpy uses BSD-3 clause licence. Arrymancer uses Apache v2 licence.

According to http://www.apache.org/legal/resolved.html#category-a BSD licenced code can be included in Apache licenced Software, and I'm pretty sure the Apache foundation has a very informed legal opinion.

**Comment by szabi** [comment]
> * What format to use as argument (strings? which convention) and how to parse it 

I think, while the implentation depth differ, the convention is *consistent* among numpy, sympy and tensorflow, so that one can speak of an industry standard. I persponally would see the choice of convention for nim/Arraymancer uncontroversial and pretty settled. I'd deviate from the established sector standard only, if the current convention has some serious limitations, such as lack of complete expressiveness.

**Comment by mratsim** [comment]
Adding another library for research: [LTensor](https://code.google.com/archive/p/ltensor/)

It uses this syntax: `Sigma(i,j)=C(i,j,k,l)*E(k,l);`

An alternative Nim syntax could be:
```Nim
einsum(Sigma, C, E):
  let i, j, k, l = EinsumIndex
  Sigma(i, j)=C(i,j,k,l)*E(k,l)
```

**Comment by giorgiop** [comment]
@mratsim what's the status of this?

**Comment by mratsim** [comment]
Hey @giorgiop, feel free to start a PR if you want to work on this.

I did not start on `einsum` at all beyond the preliminary research mentioned in this ticket.

**Comment by dimka-kopitkov** [comment]
Hi all,

I'm still experiencing some inconsistency when calculating gradients of einsum. Should it work properly in Tensorflow 1.10.0 or there are known issues with it?

Thanks

**Comment by szabi** [comment]
> An alternative Nim syntax could be:
> 
> ```nim
> einsum(Sigma, C, E):
>   let i, j, k, l = EinsumIndex
>   Sigma(i, j)=C(i,j,k,l)*E(k,l)
> ```

But wouldn't that need changes to the Nim parser and/or compiler itself (ie. not solvable in Arrymancer alone), as it gives additional semantic meaning to the variable names of the index variables? You'd have to use some kind of reflection in the overloaded `*` to be able to analyze the lexical scope *above* for the *naming* of the indices passed to the variables (`C` and `E`).

**Comment by Vindaar** [comment]
I just wrote a proof of concept for `einsum` in Arraymancer. Currently it just creates hardcoded loops around the tensor axes. I just used the syntax:
```nim
let t = einsum(c, d):
  c[i,j,k,l] * d[k,l]
```

If you think this has some merit, I can create expand on this (support less and more than 2 tensors; possibly avoid the explicit loops) and create a PR.

The code:
```nim
import arraymancer, macros, sequtils, sets 

template `^^`(s, i: untyped): untyped =
  (when i is BackwardsIndex: s.len - int(i) else: int(i))      

proc slice[T, U](n: NimNode, s: HSlice[T, U]): seq[NimNode] =
  ## returns the slice `s` of the children of `n`
  echo "s ", s.repr
  let a = n ^^ s.a
  let b = n ^^ s.b
  doAssert n.len > b, " N " & $n.len & " and b " & $b
  doAssert a >= 0
  for i in a .. b:
    result.add n[i]

proc buildLoops(rank: int, idxRes: HashSet[string],
                shapeIdent: NimNode, innerStatement: NimNode): NimNode =
  # generate the for loops
  var forLoops = nnkForStmt.newTree()
  let idxResSeq = toSeq(idxRes)
  var stmtInLoop = newNimNode(nnkNilLit)
  for i in 0 ..< rank:
    let forIdx = ident(idxResSeq[i])
    let toIdx = quote do:
      `shapeIdent`[`i`]
    var loop = nnkForStmt.newTree(
      forIdx,
      nnkInfix.newTree(
        ident"..<",
        newLit 0,
        toIdx
      )
    )
    let idxI = ident(idxResSeq[i])
    if stmtInLoop.kind == nnkNilLit:
      stmtInLoop = innerStatement
    else:
      stmtInLoop = forLoops
    loop.add stmtInLoop
    forLoops = loop
  result = forLoops

proc findAxes(idxRes: HashSet[string], tensors: seq[(NimNode, seq[NimNode])]): seq[(int, int)] =
  for idx in idxRes:
    for i, tup in tensors:
      let tIdx = tup[1].mapIt($it)
      let at = find(tIdx, $idx)
      if at >= 0:
        result.add (i, at)
        # found this index, can break from search
        break 

macro einsum(t1, t2: Tensor, stmts: untyped): untyped =
  result = newStmtList()  
  echo stmts.treeRepr
  doAssert stmts[0].kind == nnkInfix
  doAssert stmts[0][0].ident == toNimIdent"*"

  # TODO: allow nested infix w/ more than 2 tensors
  let t1C = stmts[0][1][0]
  doAssert $t1C == $t1
  let t1Idx = slice(stmts[0][1], 1 .. ^1)  
  echo t1Idx

  # let t2C = stmts[1][2][0]
  let t2C = stmts[0][2][0]
  doAssert $t2C == $t2
  let t2Idx = slice(stmts[0][2], 1 .. ^1)

  # combine tensors with their indices
  let tensors = @[(t1C, t1Idx), (t2C, t2Idx)]
  let resIdent = ident"tmp"

  # generate shape assertion statements
  for tup in tensors:
    let t = tup[0]
    let idx = tup[1]
    let nIdx = idx.len
    result.add quote do:
      doAssert `t`.rank == `nIdx`

  # generate the rank of the resulting tensor
  # TODO: avoid that `mapIt`
  let idxRes = symmetricDifference(toSet(t1Idx.mapIt($it)), toSet(t2Idx.mapIt($it)))
  let idxContr = intersection(toSet(t1Idx.mapIt($it)), toSet(t2Idx.mapIt($it)))  
  let rank = idxRes.card

  # find each axis of each tensor, which will surive
  # seq of tuples of `tensor`, `axis` pairs
  let tensorAxes = findAxes(idxRes, tensors)
  let contractionAxes = findAxes(idxContr, tensors)
  echo tensorAxes
  echo contractionAxes
  
  # generate the code to get the shape of the resulting tensor
  let shapeIdents = ident"shapes"
  result.add quote do:
    var `shapeIdents` = newSeq[int](`rank`)
  for i, ax in tensorAxes:
    let t = tensors[ax[0]][0]
    let idx = ax[1]
    result.add quote do:
      `shapeIdents`[`i`] = `t`.shape[`idx`]

  # generate the code to get the shape of the contraction
  let shapeContrIdents = ident"shapesContr"
  let rankContr = idxContr.card
  result.add quote do:
    var `shapeContrIdents` = newSeq[int](`rankContr`)
    echo "Rank contr ", `rankContr`
  for i, ax in contractionAxes:
    let t = tensors[ax[0]][0]
    let idx = ax[1]
    result.add quote do:
      `shapeContrIdents`[`i`] = `t`.shape[`idx`]      

  # generate the result tensor
  result.add quote do:
    var `resIdent` = newTensor[float](`shapeIdents`)
    echo `shapeIdents`.len
    echo `resIdent`.shape

  # generate contraction over remaining indices
  let prod = stmts[0]
  let contrRes = ident"res"
  var stmt = quote do:
    `contrRes` += `prod`
  echo stmt.repr
  var contractionLoops = newStmtList()
  contractionLoops.add quote do:
    var `contrRes`: float
  contractionLoops.add buildLoops(rankContr,
                                  idxContr,
                                  shapeContrIdents,
                                  stmt)
  # generate bracket to acceess element
  var idxBracket = nnkBracketExpr.newTree(resIdent)
  for x in idxRes:
    idxBracket.add ident(x)
  contractionLoops.add quote do:
    `idxBracket` = `contrRes`

  let forLoops = buildLoops(rank, idxRes, shapeIdents, contractionLoops)
  result.add forLoops
  
  # put everything into a block and return tmp tensor
  result = quote do:
    block:
      `result`
      `resIdent`
  echo "Result ", result.repr

let c0 = toSeq(11..34).toTensor.astype(float)
let d0 = toSeq(1..6).toTensor.astype(float)
let c = c0.reshape(2, 2, 3, 2)
let d = d0.reshape(3, 2)

let t = einsum(c, d):
  c[i,j,k,l] * d[k,l]

echo t.shape
echo t


# check against example from numpy docs
let a = toSeq(0..59).toTensor.reshape(3,4,5).astype(float)
let b = toSeq(0 .. 23).toTensor.reshape(4,3,2).astype(float)
#>>> np.einsum('ijk,jil->kl', a, b)
let ab = einsum(a, b):
  a[i,j,k] * b[j,i,l]
echo ab
  
let res = [[ 4400.0,  4730.0],
           [ 4532.0,  4874.0],
           [ 4664.0,  5018.0],
           [ 4796.0,  5162.0],
           [ 4928.0,  5306.0]].toTensor.transpose
doAssert ab == res # yay, after transposing
```

For the numpy example at the bottom the following code is created:
```nim
block:
  doAssert a.rank ==
      3
  doAssert b.rank ==
      3
  var shapes = newSeq[int](2)
  []=(shapes, 0,
      b.shape[2])
  []=(shapes, 1,
      a.shape[2])
  var shapesContr = newSeq[int](2)
  echo "Rank contr ", 2
  []=(shapesContr, 0,
      a.shape[0])
  []=(shapesContr, 1,
      a.shape[1])
  var tmp = newTensor[float](shapes)
  echo shapes.len
  echo tmp.shape
  for k in 0 ..<
      shapes[1]:
    for l in 0 ..<
        shapes[0]:
      var res: float
      for j in 0 ..<
          shapesContr[1]:
        for i in 0 ..<
            shapesContr[0]:
          res += a[i, j, k] * b[j, i, l]
      tmp[l, k] = res
  tmp

```

**Comment by mratsim** [comment]
Awesome work, this is very elegant.

Yes feel free to expand on it, the indices syntax is good. It probably only works for row-major tensors.

Regarding tests, some example from there (matrix-vector, dot product and matrix matrix multiplication) would be good for starters: https://rockt.github.io/2018/04/30/einsum

**Comment by Vindaar** [comment]
Cool, thanks! 
And thanks for the link. I'll use that as a basis for tests!
